Cogburn/bulk indexer

coreyogburn opened the pull request:
DetectionEngines were refactored to use ES's BulkIndexer decreasing import times from upwards of 20 mins down to a handful of seconds.

The improved speed introduced a problem: the way we paged large result sets fails when concurrency results in multiple detections having the same @timestamp value. By paging using the timestamp, if a page break happened in the middle of a stretch of equal @timestamp values then the next page won't begin exactly where the last page left off, leaving out between 2 and 20 detections in tests. Instead, the scrolling API is now supported. This approach to search collects all the results on ES and retains them for the client to make follow up requests to obtain the entire payload. GetAllDetections has also been improved with a call to Refresh it's indices before scrolling.

Due to the use of the bulk indexer API, the "write, no read" precautions in sync may not be necessary anymore. More research is needed, the precautions have been left in place for now.

coreyogburn committed:
Suricata + Scrolling

The bulk indexer wasn't playing well with our approach to pagination. I believe the concurrency of the import resulted in multiple documents with the same @timestamp value and we were losing some of them at the seams of pages. When expecting 50425 documents back, the paging approach was returning 50415 or 50418. Somewhere between 2 and 20 documents weren't coming back but WERE in ES. This scrolling approach is more thorough.

Suricata is using the new Bulk Indexer

All tests are passing.

coreyogburn committed:
ElastAlert

ElastAlert now uses the BulkIndexer. Tests updated.

Suricata tweaks, bug fixes, and cleanup.

coreyogburn committed:
Strelka

Realized that as soon as we observe a detection during sync we should remove it from the toDelete list. Whether that detection is updated, created, ignored, errors out, etc it doesn't matter; the detection's presence in the list means we should not delete the existing entry.

coreyogburn committed:
Final Fixes

All changes to a Detection during sync must happen before the call to ConvertObjectToDocument.

BulkIndexer now refreshes documents. GetAllDocuments refreshes the indexes before running it's query. This is to help ensure queries return accurate result sets.

Improved logging around scrolling.

All tests are passing.

coreyogburn committed:
Increase Scroll Time

Lengthen scroll time to ensure scroll can finish.

coreyogburn committed:
Fix Updates

Refactored updates to match the expected syntax. In a create you send up the raw doc, in an update you send it up wrapped in `{"doc":<doc>}`.

Turns out we were missing some errors. OnFailure seems to only provide an error object if there was an error making the request. The response object needs to be checked when there's no err. Also if the bulk indexer runs into a request level error, none of the work items get their OnFailure called, instead the BulkIndexer's OnError gets called.

Creating a bulk indexer now takes a logger which is used to print any errors that happen with the bulk indexer.

Now logging bulk inserter stats as Debug. Updated tests to account for this.

Fixed a potential race condition if multiple bulk written items fail. Prevented a panic that happens when there are multiple writes to the errMap.

Fixed a few places where PrepareForSave was clobbering the docID from previous detections.

It turns out that the backend was not preventing you from deleting a community rule. The frontend would disable the Delete button, but the backend would do it if the request was made. Useful for testing, but needed fixing. Updated with custom error message.

coreyogburn committed:
Refactor update work into ConvertObjectToDocument

Instead of converting to document and then wrapping later when we know it's an edit, ConvertObjectToDocument now takes a flag and does that work when the flag indicates. Updated references.

coreyogburn committed:
Clear Scroll when finished

Without clearing scrolls, you can get a Shard Failure error when attempting to open an additional scroll and too many are still open on the server. We don't have issues with all the engines scrolling at once, we were seeing issues if all the engines sync repeatedly.

coreyogburn committed:
Refactor BulkUpdate to use BulkIndexer

Although the bulk update is incredibly fast, we still need to sync 1 to 3 engines after it's done so the handler is still split into 2 parts: a sync part that verifies you're not trying to delete a community rule, and an async part that makes the changes, does the necessary syncs, and broadcasts the results.

The UI now displays a loading circle during the call to start a bulk action.

Fixed a bug in elastalert sync that was preventing audit entries from getting into ES.

Removed the UpdateDetectionField function from the DetectionStore interface. It was only used in Bulk Updates and has been completely replaced.

TruncateMap was made generic.

coreyogburn committed:
Cleanup

Refactored EventScrollCriteria to only the necessary parts. The criteria are used to make a large request with no metrics, no date ranges, no sorting, no search after.

The converter cleans up to just the query and the size which is hardcoded at the ES search limit. A few tests have been added.

jertel reviewed (dismissed):
Approving with request for minor changes in a follow-up PR.

coreyogburn committed:
Changes from Code Review

In bulkUpdateDetectionAsync, all the commented out web.Respond statements have been replaced with log statements. Audit log statements are marked as such.

AuditInfo moved to model namespace.

The converter now can be configured for what the maxScrollSize is instead of assuming 10K.

The bulk indexer now will use a configurable number of workers.

ConvertObjectToDocument no longer uses the kind parameter to help fill out the index return value.
